### DISCUSSION TITLE
StepCI feedback

### DISCUSSION THREAD

**Discussion by notmedia** [stepci/stepci#80] [Ideas]
### Concepts
https://docs.stepci.com/guides/concepts.html

1. Can I use “allow failing” option for step in test?

### HTTP APIs
https://docs.stepci.com/guides/testing-http.html

1. Usually in tests we have pre-test setup logic with authentication which generates to us for example access token. It will be better if I can do it here - generate access token and then use it next steps. — I see here that we can use one request and then use the data from this request in another steps, so maybe it’s already solved. 
2. The default approach for authentication with access token is also having a refresh token. Access token provide us permissions to use API and has TTL for a small time (5 mins for example), refresh token provide us permissions to renew access token and has TTL for a long time. So it would be grate if we have an ability to obtain new access token by requesting it with refresh token if access token expired. If we have test suite with long time execution access token may expire for some time, and this will be helpful if we can automatically renew it.

### GraphQL APIs
https://docs.stepci.com/guides/testing-graphql.html

1. Add subscriptions support I think right now it’s not supported yet.

### gRPC APIs
https://docs.stepci.com/guides/testing-grpc.html  

Disclaimer - maybe for all of this features you can use my core library - https://github.com/getezy/ezy/tree/master/src/core

1. Support gRPC-Web protocol - I know that right now we can query gRPC-Web endpoints using HTTP tests, but this is not actually fits user need. In gRPC-Web protocol they define routes for each methods and it hard to remember and type them. Also they are sending serialised data, and you need to deserialise it by provided Protobuf files.
2. Support streams for gRPC and gRPC-Web. For all streams we need to have the ability to chain messages and have options for waiting for response from the server. Also support cancelation and stoping stream from client.
  **Example of complex case:**
   Bidirectional streaming → Init request → Send message from client 2 times → Wait for response from server and check data -> Repeat 4 times -> Stop client stream -> Wait for ending server stream. We can discuss about all cases later.
3. Auth in gRPC often uses access token in metadata, we need to have an ability like in HTTP for query an auth method → obtain token → pass it to all metadata requests in the test suite.
4. Can’t send Unary request for this server - https://github.com/getezy/ezy/blob/master/__tests__/simple-service/proto/simple.proto

With this test suite:
```yaml
version: "1.1"
name: My Workflow
env:
  host: 0.0.0.0:4000
tests:
  example:
    name: Example test
    steps:
      - name: Example step
        grpc:
          proto:
            - simple.proto
          host: ${{env.host}}
          service: simple_package.v1.SimpleService
          method: Unary
          data:
            id: test
```

Getting an error: `12 UNIMPLEMENTED: The server does not implement the method /simple_package.v1/Unary`

I think there is a bug in splitting service name, I just want to say that in gRPC we often use tags like `.v1` in package name for version history.

Tried to use package without `.v1`:
```
syntax = "proto3";

package simple_package;
...
```

Test suite:
```yaml
...
          service: simple_package.SimpleService
          method: Unary
...
```

Getting an error: `12 UNIMPLEMENTED: The server does not implement the method /simple_package.SimpleService/Unary`

Tried to use it without package by removing `package` definition from proto file.

Test suite:
```yaml
...
          service: SimpleService
          method: Unary
...
```

Getting an error: `12 UNIMPLEMENTED: The server does not implement the method /SimpleService.undefined/Unary
`

5. Can I use absolute paths for my protobuf files? Use case is simple - tests and protobuf files may be separated by different projects/folders.
6. I have a question, only after solving 4 point can test it, just imagine:
You have `main.proto` file with this definition:
```
syntax = "proto3";

package simple_package.v1;

import "types.proto";

service SimpleService {
  rpc Unary(SimpleMessage) returns (SimpleMessage);
}
```

And your `types.proto` file:
```
syntax = "proto3";

message SimpleMessage {
  string id = 1;
}
```
If we define `main.proto` in test suite, shall it work and use imported message from `types.proto`?

### Load Testing
https://docs.stepci.com/guides/load-testing.html

1. Can I use load testing for gRPC services?

### Test Data
https://docs.stepci.com/guides/using-test-data.html

1. “You can load test data from a csv file/string to be used in tests.” - I don’t understand how to use string format?

### Features

1. Support matrix for environments and data, like in GitHub actions - https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs - useful when I want to test same API endpoint with different values, this allows us to prevent duplication of test suites.
2. UI Builder for test suites is a killer feature I think.

#### Reference → Workflow Syntax 
1. Divide this page by sections and use links across your docs to related section.

#### Testing HTTP APIs 
1. Add more info for JSONPath expressions, or a link to this library.
2. Add more info for JSON Schema options.
3. Add more info how to use CSS selectors (elements, tags, classes, attributes) and how to check for example the value from this selector. Or if we have a list of items and want to check items count, how I can do it?
4. Add more info for Performance checks, for dumb users say that this values are in milliseconds.

#### Testing gRPC APIs

1. Add Protobuf file that you use in test example.

**Comment by mishushakov** [comment]
Hello hello, thanks for the feedback

> Can I use “allow failing” option for step in test?

Yes, I have highlighted it in the documentation

> Usually in tests we have pre-test setup logic with authentication which generates to us for example access token. It will be better if I can do it here - generate access token and then use it next steps

You can capture data from previous steps and use it in consequent steps (in documentation)

> The default approach for authentication with access token is also having a refresh token

You should create a separate token for CI/CD that does not expire

> Add (GraphQL) subscriptions support I think right now it’s not supported yet

Realtime APIs are on backlog, but I still need to think how it will be implemented

> Support gRPC-Web protocol

I don't think makes much sense. Typically gRPC-Web Gateway is just a proxy to the gRPC API?

> Support streams for gRPC and gRPC-Web

Client and server streaming is already supported by cool-grpc, but not Bidirectional Streaming and not for gRPC-Web

> Auth in gRPC often uses access token in metadata, we need to have an ability like in HTTP for query an auth method

You can capture data from a previous step and send it in the metadata field in the next step

> I think there is a bug in splitting service name

Should be fixed now. Can you try again?

> Can I use absolute paths for my protobuf files? Use case is simple

Yes, you should be able to. Imports are not supported though

> Can I use load testing for gRPC services?

Yes, I have highlighted it in the documentation

> “You can load test data from a csv file/string to be used in tests.” - I don’t understand how to use string format?

https://docs.stepci.com/reference/workflow-syntax.html#tests-test-testdata

Will take all your feature and documentations suggestions into consideration

I have another question about the gRPC library you use in ezy. I can see us using it as well. Would you be able to publish it as a package on NPM and include documentation how to use it? We can wrap it in our cool-grpc repository for easier access

Thank you!

**Comment by notmedia** [comment]
Hi! 
> You should create a separate token for CI/CD that does not expire

Ah, I thought that it useful when the user wants to create integration test. I don't know is the CI/CD token without expiration is a best practice. Recurrent query by condition I think could be a great feature.

> I don't think makes much sense. Typically gRPC-Web Gateway is just a proxy to the gRPC API?

This is not just a proxy, it transmits serialised data by protobuf definitions and deserialise data on income. Also it forms HTTP paths for this proxy.

> Should be fixed now. Can you try again?

Yep I will try and return back

> Yes, you should be able to. Imports are not supported though

For gRPC imports are necessary. In real projects usually protos are divided by files

> I have another question about the gRPC library you use in ezy. I can see us using it as well. Would you be able to publish it as a package on NPM and include documentation how to use it? We can wrap it in our cool-grpc repository for easier access

I think yes, but I need some time. I want to rewrite it and add full test coverage.
